File: ros2_ws/src/blindspot_ros/blindspot_ros/fake_detections.py
"""Synthetic detections, so the guard node can be exercised without a camera.

Walks a virtual camera through three regimes and publishes the resulting
marker centres, which is enough to see the guard change its mind:

    healthy    all 6 markers on the ring          -> partition ok
    occluded   only 2 markers survive             -> partition dropped
    collapsed  all 6 markers, target near a line  -> partition dropped

    ros2 run blindspot_ros fake_detections
"""

import numpy as np
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSDurabilityPolicy, QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import CameraInfo
from vision_msgs.msg import (BoundingBox2D, Detection2D, Detection2DArray,
                             ObjectHypothesisWithPose)

WIDTH, HEIGHT, FOVY = 1920, 1440, 45.0
PHASE = 40                      # frames per regime


def ring(r=0.05, n=6):
    a = np.linspace(0, 2 * np.pi, n + 1)[:-1]
    return np.stack([r * np.cos(a), r * np.sin(a), np.zeros(n)], axis=1)


def project(points, cTo, fx, fy, cx, cy):
    P = np.hstack([points, np.ones((len(points), 1))])
    P_c = (cTo @ P.T).T[:, :3]
    s = np.stack([P_c[:, 0] / P_c[:, 2], P_c[:, 1] / P_c[:, 2]], axis=1)
    return s * np.array([fx, fy]) + np.array([cx, cy])


class Fake(Node):

    def __init__(self):
        super().__init__("fake_detections")
        fy = (HEIGHT / 2.0) / np.tan(np.deg2rad(FOVY) / 2.0)
        self.K = (fy, fy, (WIDTH - 1) / 2.0, (HEIGHT - 1) / 2.0)
        latched = QoSProfile(depth=1,
                             reliability=ReliabilityPolicy.RELIABLE,
                             durability=QoSDurabilityPolicy.TRANSIENT_LOCAL)
        self.info_pub = self.create_publisher(CameraInfo, "camera_info",
                                              latched)
        self.det_pub = self.create_publisher(Detection2DArray, "detections",
                                             10)
        self.k = 0
        self.publish_info()
        self.create_timer(0.1, self.tick)

    def publish_info(self):
        m = CameraInfo()
        m.width, m.height = WIDTH, HEIGHT
        fx, fy, cx, cy = self.K
        m.k = [fx, 0.0, cx, 0.0, fy, cy, 0.0, 0.0, 1.0]
        m.header.frame_id = "camera"
        self.info_pub.publish(m)

    def tick(self):
        phase = (self.k // PHASE) % 3
        pts = ring()
        keep = list(range(6))
        if phase == 1:
            keep = [0, 3]
        elif phase == 2:
            pts = ring()
            pts[:, 1] *= 0.02
        cTo = np.eye(4)
        cTo[:3, 3] = [0.02 * np.sin(self.k * 0.05), -0.01, 0.62]
        uv = project(pts, cTo, *self.K)

        msg = Detection2DArray()
        msg.header.frame_id = "camera"
        msg.header.stamp = self.get_clock().now().to_msg()
        # deliberately shuffled: a detector reports whatever order it finds,
        # and the guard node must recover the order from the ids.
        order = list(keep)
        rng = np.random.default_rng(self.k)
        rng.shuffle(order)
        for i in order:
            d = Detection2D()
            d.id = str(i)
            b = BoundingBox2D()
            b.center.position.x = float(uv[i, 0])
            b.center.position.y = float(uv[i, 1])
            b.size_x, b.size_y = 40.0, 40.0
            d.bbox = b
            h = ObjectHypothesisWithPose()
            h.hypothesis.class_id = str(i)
            h.hypothesis.score = 1.0
            d.results.append(h)
            msg.detections.append(d)
        self.det_pub.publish(msg)

        if self.k % PHASE == 0:
            self.get_logger().info(
                "phase %d: %s" % (phase, ["healthy, 6 markers",
                                          "occluded, 2 markers",
                                          "collapsed target, 6 markers"][phase]))
        self.k += 1


def main(argv=None):
    rclpy.init(args=argv)
    n = Fake()
    try:
        rclpy.spin(n)
    except KeyboardInterrupt:
        pass
    finally:
        n.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/blindspot_cpp/src/fake_detections.cpp
// Synthetic detections, so the guard node can be exercised without a camera.
//
// Walks a virtual camera through three regimes and publishes the resulting
// marker centres, which is enough to see the guard change its mind:
//
//     healthy    all 6 markers on the ring          -> partition ok
//     occluded   only 2 markers survive             -> partition dropped
//     collapsed  all 6 markers, target near a line  -> partition dropped
//
//     ros2 run blindspot_cpp fake_detections
#include <algorithm>
#include <chrono>
#include <cmath>
#include <memory>
#include <random>
#include <string>
#include <vector>

#include <Eigen/Core>

#include <rclcpp/rclcpp.hpp>

#include <sensor_msgs/msg/camera_info.hpp>
#include <vision_msgs/msg/detection2_d_array.hpp>

#include "blindspot/units.hpp"

namespace
{

constexpr int WIDTH = 1920;
constexpr int HEIGHT = 1440;
constexpr double FOVY = 45.0;
constexpr int PHASE = 40;               // frames per regime

Eigen::MatrixX3d ring(double r = 0.05, int n = 6)
{
  Eigen::MatrixX3d p(n, 3);
  for (int i = 0; i < n; ++i) {
    const double a = 2.0 * M_PI * i / n;
    p(i, 0) = r * std::cos(a);
    p(i, 1) = r * std::sin(a);
    p(i, 2) = 0.0;
  }
  return p;
}

class Fake : public rclcpp::Node
{
public:
  Fake()
  : Node("fake_detections")
  {
    k_intr_ = blindspot::intrinsics_from_fov(WIDTH, HEIGHT, FOVY);
    const auto latched = rclcpp::QoS(1).reliable().transient_local();
    info_pub_ = create_publisher<sensor_msgs::msg::CameraInfo>("camera_info", latched);
    det_pub_ = create_publisher<vision_msgs::msg::Detection2DArray>("detections", 10);
    publish_info();
    timer_ = create_wall_timer(std::chrono::milliseconds(100), [this]() {tick();});
  }

private:
  void publish_info()
  {
    sensor_msgs::msg::CameraInfo m;
    m.width = WIDTH;
    m.height = HEIGHT;
    m.k = {k_intr_.fx, 0.0, k_intr_.cx, 0.0, k_intr_.fy, k_intr_.cy, 0.0, 0.0, 1.0};
    m.header.frame_id = "camera";
    info_pub_->publish(m);
  }

  void tick()
  {
    const int phase = (k_ / PHASE) % 3;
    Eigen::MatrixX3d pts = ring();
    std::vector<int> keep{0, 1, 2, 3, 4, 5};
    if (phase == 1) {
      keep = {0, 3};
    } else if (phase == 2) {
      pts.col(1) *= 0.02;
    }
    Eigen::Vector3d t(0.02 * std::sin(k_ * 0.05), -0.01, 0.62);

    Eigen::MatrixX2d uv(pts.rows(), 2);
    for (Eigen::Index i = 0; i < pts.rows(); ++i) {
      const Eigen::Vector3d p = pts.row(i).transpose() + t;
      uv(i, 0) = (p(0) / p(2)) * k_intr_.fx + k_intr_.cx;
      uv(i, 1) = (p(1) / p(2)) * k_intr_.fy + k_intr_.cy;
    }

    vision_msgs::msg::Detection2DArray msg;
    msg.header.frame_id = "camera";
    msg.header.stamp = now();
    // deliberately shuffled: a detector reports whatever order it finds, and
    // the guard node must recover the order from the ids.
    std::vector<int> order = keep;
    std::mt19937 rng(static_cast<unsigned int>(k_));
    std::shuffle(order.begin(), order.end(), rng);
    for (const int i : order) {
      vision_msgs::msg::Detection2D d;
      d.id = std::to_string(i);
      d.bbox.center.position.x = uv(i, 0);
      d.bbox.center.position.y = uv(i, 1);
      d.bbox.size_x = 40.0;
      d.bbox.size_y = 40.0;
      vision_msgs::msg::ObjectHypothesisWithPose h;
      h.hypothesis.class_id = std::to_string(i);
      h.hypothesis.score = 1.0;
      d.results.push_back(h);
      msg.detections.push_back(d);
    }
    det_pub_->publish(msg);

    if (k_ % PHASE == 0) {
      static const char * kNames[3] = {
        "healthy, 6 markers", "occluded, 2 markers", "collapsed target, 6 markers"};
      RCLCPP_INFO(get_logger(), "phase %d: %s", phase, kNames[phase]);
    }
    ++k_;
  }

  blindspot::Intrinsics k_intr_;
  int k_{0};
  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr info_pub_;
  rclcpp::Publisher<vision_msgs::msg::Detection2DArray>::SharedPtr det_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

}  // namespace

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Fake>());
  rclcpp::shutdown();
  return 0;
}
